# Data Consistency Checks Products and orders

# Importing Libraries and datasets

# Reading Tasks

## Creating a Dataframe

## Consistency Checks on Products

# Start of Exercise

## Consistency Check on Orders

# Export Data sets

--------

# Importing Libraries and datasets

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#creating pathway for dataframes
path=os.path.join(r'C:\Users\colmu\Desktop\Career Foundry\Achievement 4\Instacart Basket Analysis\Data')

In [4]:
#import products.csv
df_prods= pd.read_csv(os.path.join(path,'Original Data', 'product.csv'),index_col=False)

In [5]:
#import orders_wrangled.csv
df_ords= pd.read_csv(os.path.join(path,'Prepared Data', 'orders_wrangled.csv'))

# Reading Tasks

## Creating a Dataframe

In [6]:
# Create a dataframe
df_test= pd.DataFrame()

In [7]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

#check
df_test.head()

mix
0     a
1     b
2     1
3  True

## Consistency Checks on Products

In [8]:
# Describe check on products df
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [9]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
#checking for missing data
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
#new df for product Nan values
df_nan=df_prods[df_prods['product_name'].isnull()==True]

#check
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
#new products df with no missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

#shape check
df_prods_clean.shape

(49677, 5)

In [18]:
#new df for finding duplicates
df_dups=df_prods_clean[df_prods_clean.duplicated()]

#check
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
#new df with no duplicates
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

# Start of Exercise

## Consistency Check on Orders

In [19]:
#adding code to get rid of exponents
pd.options.display.float_format = '{:.2f}'.format

In [22]:
#Task 2 investigating orders_df - no concerns found
df_ords.describe()

order_id    user_id  clients_order_number  orders_day_of_week  \
count 3421083.00 3421083.00            3421083.00          3421083.00   
mean  1710542.00  102978.21                 17.15                2.78   
std    987581.74   59533.72                 17.73                2.05   
min         1.00       1.00                  1.00                0.00   
25%    855271.50   51394.00                  5.00                1.00   
50%   1710542.00  102689.00                 11.00                3.00   
75%   2565812.50  154385.00                 23.00                5.00   
max   3421083.00  206209.00                100.00                6.00   

       order_hour_of_day  days_since_prior_order  
count         3421083.00              3214874.00  
mean               13.45                   11.11  
std                 4.23                    9.21  
min                 0.00                    0.00  
25%                10.00                    4.00  
50%                13.00                    7.00  
75%                16.00                   15.00  
max                23.00                   30.00

In [23]:
#Task 3 Checking for mixed type data - no issues found
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [24]:
#Task 5 Missing Values check
df_ords.isnull().sum()

order_id                       0
user_id                        0
clients_order_number           0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## Task 5 days_since_prior_order has 206209 NAN values. This variable is capped at 30 days and will blow in NaN values for users who have not placed an instacart order in over 30 days.

## Task 6 A flag variable in a new column should be created marking active customers(orders placed in under 31 days) and inactive customers (31 days or over) once I learn how to do this. 

# Export Data sets

In [20]:
#exporting cleaned products df
df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data', 'products_checked.csv'), index=False)

In [ ]:
# Task 9 - I did nothing to my orders_DF but I will save a new one for less confusion
df_ords.to_csv(os.path.join(path,'Prepared Data', 'orders_checked.csv'), index=False)